lemur/api_keys/views.py

```diff
@@ -0,0 +1,558 @@
+        access_key = service.get(aid)
+        if access_key is None:
+            return dict(message="You are not authorized to update this token!"), 403
+        if access_key.user_id != uid:
+            return dict(message="You are not authorized to update this token!"), 403
+        service.update(access_key, name=data['name'], revoked=data['revoked'], ttl=data['ttl'])
+        return dict(jwt=create_token(access_key.user_id, access_key.id, access_key.ttl))
+
+    def delete(self, uid, aid):
+        """
+        .. http:delete:: /users/1/api_keys/1
+
+           deletes one api key
+
+           **Example request**:
+
+           .. sourcecode:: http
+
+              DELETE /users/1/api_keys/1 HTTP/1.1
+              Host: example.com
+              Accept: application/json, text/javascript
+
+           **Example response**:
+
+           .. sourcecode:: http
+
+              HTTP/1.1 200 OK
+              Vary: Accept
+              Content-Type: text/javascript
+
+              {
+                  "result": true
+              }
+
+           :reqheader Authorization: OAuth token to authenticate
+           :statuscode 200: no error
+           :statuscode 403: unauthenticated
+        """
+        if uid != g.current_user.id:
+            if not ApiKeyCreatorPermission().can():
+                return dict(message="You are not authorized to view this token!"), 403
+        access_key = service.get(aid)
+        if access_key is None:
+            return dict(message="You are not authorized to delete this token!"), 403
+        if access_key.user_id != uid:
+            return dict(message="You are not authorized to delete this token!"), 403
+        service.delete(access_key)
+        return {'result': True}
+
+
+class ApiKeysDescribed(AuthenticatedResource):
+    def __init__(self):
+        self.reqparse = reqparse.RequestParser()
+        super(ApiKeysDescribed, self).__init__()
+
+    @validate_schema(None, api_key_described_output_schema)
+    def get(self, aid):
+        """
+        .. http:get:: /api_keys/1/described
+
+           Fetch one api key
+
+           **Example request**:
+
+           .. sourcecode:: http
+
+              GET /api_keys/1 HTTP/1.1
+              Host: example.com
+              Accept: application/json, text/javascript
+
+           **Example response**:
+
+           .. sourcecode:: http
+
+              HTTP/1.1 200 OK
+              Vary: Accept
+              Content-Type: text/javascript
+
+              {
+                  "id": 2,
+                  "name": "hoi",
+                  "user_id": 2,
+                  "ttl": -1,
+                  "issued_at": 1222222,
+                  "revoked": false
+              }
+
+           :reqheader Authorization: OAuth token to authenticate
+           :statuscode 200: no error
+           :statuscode 403: unauthenticated
+        """
+        access_key = service.get(aid)
+        if access_key is None:
```

Comment:
I would consider reversing these two, throw a 403 if the user truly doesn't have permission then throw a 404 if the user has permission but the API key does not exist.

Reply:
Hmm, that's a good idea. I just took the 403 always to not leak info (like S3, etc.), but I like throwing a 404 if they do have permission, more.

Reply:
Actually while looking at doing this, there isn't a good way of doing this. Because we only get the access key id passed in. So we can't determine if it belongs to the user or not (and thus has permission) without the access key existing.

Reply:
Could we do something similar to what we do with fetching private keys? 

https://github.com/Netflix/lemur/blob/master/lemur/certificates/views.py#L434

Here we determine first if the Id exists and then if you have access. I suppose you could leak information about the aid itself, but aren't those sequentially issued anyway?

Reply:
Yea that could be done, but as you mentioned it would leak the access key id. However, as you also mentioned those are also issued sequentially. Technically a legitimate user should never see this message so I aired on the side of safety.

I can change it though if you'd like.

Reply:
Yeah, let's go ahead and make that change. I'd like to maintain the parity we have with viewing private keys.

Reply:
Alright, that change has been made. Non-existent keys will now show 404.